File: src/bus.rs
// bus.rs

use crate::{Memory, Audio, DiskController, Keyboard, Mouse, Overlay};

use std::sync::{Arc, Mutex};
use std::io::{Write, stdout};

pub struct Bus {
    pub memory: Box<dyn Memory>,

    pub disk_controller: DiskController,

    pub audio: Arc<Mutex<Audio>>,

    pub keyboard: Arc<Mutex<Keyboard>>,
    pub mouse: Arc<Mutex<Mouse>>,

    pub overlays: Arc<Mutex<Vec<Overlay>>>,
}

impl Bus {
    pub fn read_io(&mut self, port: u32) -> u32 {
        match port {
            0x80000000..=0x8000031F => { // overlay port
                let overlay_lock = self.overlays.lock().unwrap();
                let overlay_number = (port & 0x000000FF) as usize;
                let setting = (port & 0x0000FF00) >> 8;

                match setting {
                    0x00 => {
                        // we're reading the position of this overlay
                        let x = overlay_lock[overlay_number].x as u32;
                        let y = overlay_lock[overlay_number].y as u32;
                        (y << 16) | x
                    }
                    0x01 => {
                        // we're reading the size of this overlay
                        let width = overlay_lock[overlay_number].width as u32;
                        let height = overlay_lock[overlay_number].height as u32;
                        (height << 16) | width
                    }
                    0x02 => {
                        // we're reading the framebuffer pointer of this overlay
                        overlay_lock[overlay_number].framebuffer_pointer
                    }
                    0x03 => {
                        // we're reading the enable status of this overlay
                        overlay_lock[overlay_number].enabled as u32
                    }
                    _ => panic!("invalid overlay control port"),
                }
            }
            0x80000400..=0x80000401 => { // mouse port
                let setting = (port & 0x000000FF) as u8;
                match setting {
                    0x00 => {
                        // we're reading the button states
                        let mut byte: u8 = 0x00;
                        let mouse_lock = self.mouse.lock().expect("failed to lock the mouse mutex");
                        if mouse_lock.clicked {
                            byte |= 0b001;
                        }
                        if mouse_lock.released {
                            byte |= 0b010;
                        }
                        if mouse_lock.held {
                            byte |= 0b100;
                        } else {
                            byte &= !0b100;
                        }
                        byte as u32
                    }
                    0x01 => {
                        // we're reading the position
                        let mouse_lock = self.mouse.lock().expect("failed to lock the mouse mutex");
                        let x = mouse_lock.x as u32;
                        let y = mouse_lock.y as u32;
                        (y << 16) | x
                    }
                    _ => panic!("invalid mouse control port"),
                }
            }
            0x80000500 => { // keyboard port
                let mut keyboard_lock = self.keyboard.lock().expect("failed to lock the keyboard mutex");
                keyboard_lock.pop() as u32
            }
            0x80001000..=0x80002003 => { // disk controller port
                let id = port as u8;
                let operation = (port & 0x0000F000) >> 8;

                match operation {
                    0x10 => {
                        // we're reading the current insert state of the specified disk id
                        if id > 3 {
                            panic!("invalid disk ID");
                        }
                        match &self.disk_controller.disk[id as usize] {
                            Some(disk) => disk.size() as u32, // return size if this disk is inserted
                            None => 0, // return 0 if this disk is not inserted
                        }
                    }
                    0x20 => {
                        // we're getting the location of the memory sector buffer
                        self.disk_controller.buffer_pointer as u32
                    }
                    _ => panic!("invalid disk controller port"),
                }
            }
            _ => 0,
        }
    }
    pub fn write_io(&mut self, port: u32, word: u32) {
        match port {
            0x00000000 => { // terminal output port
                print!("{}", word as u8 as char);
                stdout().flush().expect("could not flush stdout");
            }
            0x80000000..=0x8000031F => { // overlay port
                let mut overlay_lock = self.overlays.lock().unwrap();
                let overlay_number = (port & 0x000000FF) as usize;
                let setting = (port & 0x0000FF00) >> 8;

                match setting {
                    0x00 => {
                        // we're setting the position of this overlay
                        let x = (word & 0x0000FFFF) as usize;
                        let y = ((word & 0xFFFF0000) >> 16) as usize;
                        overlay_lock[overlay_number].x = x;
                        overlay_lock[overlay_number].y = y;
                    }
                    0x01 => {
                        // we're setting the size of this overlay
                        let width = (word & 0x0000FFFF) as usize;
                        let height = ((word & 0xFFFF0000) >> 16) as usize;
                        overlay_lock[overlay_number].width = width;
                        overlay_lock[overlay_number].height = height;
                    }
                    0x02 => {
                        // we're setting the framebuffer pointer of this overlay
                        overlay_lock[overlay_number].framebuffer_pointer = word;
                    }
                    0x03 => {
                        // we're setting the enable status of this overlay
                        overlay_lock[overlay_number].enabled = word != 0;
                    }
                    _ => panic!("invalid overlay control port"),
                }
            }
            0x80000400..=0x80000401 => { // mouse port
                let setting = (port & 0x000000FF) as u8;
                match setting {
                    0x00 => {
                        // we're setting the button states
                        let mut mouse_lock = self.mouse.lock().expect("failed to lock the mouse mutex");
                        mouse_lock.clicked = word & (1 << 0) != 0;
                        mouse_lock.released = word & (1 << 1) != 0;
                        mouse_lock.held = word & (1 << 2) != 0;
                    }
                    0x01 => {
                        // we're setting the position
                        let mut mouse_lock = self.mouse.lock().expect("failed to lock the mouse mutex");
                        let x = (word & 0x0000FFFF) as u16;
                        let y = ((word & 0xFFFF0000) >> 16) as u16;
                        mouse_lock.x = x;
                        mouse_lock.y = y;
                    }
                    _ => panic!("invalid mouse control port"),
                }
            }
            0x80000600 => { // audio port
                let mut audio_lock = self.audio.lock().unwrap();
                audio_lock.playing = word != 0;
                audio_lock.current_buffer_is_0 = true;
            }
            0x80001000..=0x80005003 => { // disk controller port
                let id = port as u8;
                let operation = (port & 0x0000F000) >> 8;

                match operation {
                    0x10 => {
                        // we're requesting a disk to be inserted to the specified disk id
                        if id > 3 {
                            panic!("invalid disk ID");
                        }
                        let file = self.disk_controller.select_file();
                        match file {
                            Some(file) => self.disk_controller.insert(file, id),
                            None => {},
                        };
                    }
                    0x20 => {
                        // we're setting the location of the memory sector buffer
                        self.disk_controller.buffer_pointer = word as usize;
                    }
                    0x30 => {
                        // we're reading the specified sector of the specified disk id into the memory sector buffer
                        if id > 3 {
                            panic!("invalid disk ID");
                        }
                        self.disk_controller.set_current_sector(id, word);
                        self.disk_controller.read_into_memory(id, self.memory.ram());
                    }
                    0x40 => {
                        // we're writing the specified sector to the specified disk id from the memory sector buffer
                        if id > 3 {
                            panic!("invalid disk ID");
                        }
                        self.disk_controller.set_current_sector(id, word);
                        self.disk_controller.write_from_memory(id, self.memory.ram());
                    }
                    0x50 => {
                        // we're requesting a disk to be removed from the specified disk id
                        if id > 3 {
                            panic!("invalid disk ID");
                        }
                        self.disk_controller.remove(id);
                    }
                    _ => panic!("invalid disk controller port"),
                }
            }
            _ => return,
        }
    }
}

impl fox32core::Bus for Bus {
    fn io_read(&mut self, port: u32) -> Option<u32> {
        Some(self.read_io(port))
    }

    fn io_write(&mut self, port: u32, value: u32) -> Option<()> {
        self.write_io(port, value);
        Some(())
    }
}

File: src/bus.c
#include <SDL2/SDL.h>
#include <getopt.h>
#include <math.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <time.h>

#include "bus.h"
#include "cpu.h"
#include "disk.h"
#include "framebuffer.h"
#include "keyboard.h"
#include "mouse.h"
#include "serial.h"
#include "sound.h"

bool bus_requests_exit = false;

extern time_t rtc_time;
extern uint32_t rtc_uptime;

extern fox32_vm_t vm;
extern disk_controller_t disk_controller;
extern mouse_t mouse;
extern sound_t snd;

int bus_io_read(void *user, uint32_t *value, uint32_t port) {
    (void) user;
    switch (port) {
#ifndef WINDOWS
        case 0x00000000: { // serial port
            *value = serial_get();
            break;
        };
#endif
        case 0x80000000 ... 0x8000031F: { // overlay port
            uint8_t overlay_number = port & 0x000000FF;
            uint8_t setting = (port & 0x0000FF00) >> 8;
            switch (setting) {
                case 0x00: {
                    // overlay position
                    uint32_t x = overlay_get(overlay_number)->x;
                    uint32_t y = overlay_get(overlay_number)->y;
                    *value = (y << 16) | x;
                    break;
                };
                case 0x01: {
                    // overlay size
                    uint32_t width = overlay_get(overlay_number)->width;
                    uint32_t height = overlay_get(overlay_number)->height;
                    *value = (height << 16) | width;
                    break;
                };
                case 0x02: {
                    // overlay framebuffer pointer
                    *value = overlay_get(overlay_number)->pointer;
                    break;
                };
                case 0x03: {
                    // overlay enable status
                    *value = overlay_get(overlay_number)->enabled ? 1 : 0;
                    break;
                };
            }

            break;
        };

        case 0x80000400 ... 0x80000401: { // mouse port
            uint8_t setting = port & 0x000000FF;
            switch (setting) {
                case 0x00: {
                    // button states
                    if (mouse.clicked) *value |= 0b001;
                    if (mouse.released) *value |= 0b010;
                    if (mouse.held) *value |= 0b100; else *value &= ~(0b100);
                    break;
                };
                case 0x01: {
                    // position
                    *value = (mouse.y << 16) | mouse.x;
                    break;
                };
            }

            break;
        };

        case 0x80000500: {
            *value = (uint32_t) key_take();

            break;
        }
        case 0x80000600 ... 0x80000680: { // audio port
            size_t id = port & 0xFF;
            uint8_t channel = (id & 0x30) >> 4;
            uint8_t reg = (id & 0x0F);
            switch (id) {
                case 0x80: {
                    // AUDBASE
                    *value = snd.base;
                    break;
                }
            }
            switch (reg) {
                case 0x0: {
                    // AUDxPOS
                    *value = snd.channel[channel].position;
                    break;
                }
                case 0x1: {
                    // AUDxDAT
                    *value = snd.channel[channel].data;
                    break;
                }
                case 0x4: {
                    // AUDxRATE
                    *value = snd.channel[channel].accumulator;
                    break;
                }
                case 0x5: {
                    // AUDxCONTROL
                    *value = snd.channel[channel].volume | (snd.channel[channel].loop << 7) | (snd.channel[channel].enable << 8) | (snd.channel[channel].bits16 << 9);
                    break;
                }
                case 0x6: {
                    // AUDxPAN
                    *value = snd.channel[channel].right_volume | (snd.channel[channel].left_volume << 8);
                    break;
                }
            }
            break;
        }

        case 0x80000700 ... 0x80000707: { // RTC port
            uint8_t setting = port & 0x000000FF;
            struct tm *now = localtime(&rtc_time);
            switch (setting) {
                case 0x00: { // year
                    *value = now->tm_year + 1900;
                    break;
                }
                case 0x01: { // month
                    *value = now->tm_mon + 1;
                    break;
                }
                case 0x02: { // day
                    *value = now->tm_mday;
                    break;
                }
                case 0x03: { // hour
                    *value = now->tm_hour;
                    break;
                }
                case 0x04: { // minute
                    *value = now->tm_min;
                    break;
                }
                case 0x05: { // second
                    *value = now->tm_sec;
                    break;
                }
                case 0x06: { // ms since startup
                    *value = rtc_uptime;
                    break;
                }
                case 0x07: { // daylight savings time active
                    *value = now->tm_isdst;
                    break;
                }
            }

            break;
        }

        case 0x80001000 ... 0x80002003: { // disk controller port
            size_t id = port & 0xFF;
            uint8_t operation = (port & 0x0000F000) >> 8;
            switch (operation) {
                case 0x10: {
                    // current insert state of specified disk id
                    // size will be zero if disk isn't inserted
                    *value = get_disk_size(id);
                    break;
                };
                case 0x20: {
                    // current buffer pointer
                    *value = disk_controller.buffer_pointer;
                    break;
                };
            }

            break;
        };
    }

    return 0;
}

int bus_io_write(void *user, uint32_t value, uint32_t port) {
    (void) user;
    switch (port) {
        case 0x00000000: { // serial port
            serial_put(value);
            break;
        };

        case 0x80000000 ... 0x8000031F: { // overlay port
            uint8_t overlay_number = (port & 0x000000FF);
            uint8_t setting = (port & 0x0000FF00) >> 8;
            switch (setting) {
                case 0x00: {
                    // overlay position
                    uint32_t x = value & 0x0000FFFF;
                    uint32_t y = (value & 0xFFFF0000) >> 16;
                    overlay_get(overlay_number)->x = x;
                    overlay_get(overlay_number)->y = y;
                    break;
                };
                case 0x01: {
                    // overlay size
                    uint32_t width = value & 0x0000FFFF;
                    uint32_t height = (value & 0xFFFF0000) >> 16;
                    overlay_get(overlay_number)->width = width;
                    overlay_get(overlay_number)->height = height;
                    break;
                };
                case 0x02: {
                    // overlay framebuffer pointer
                    overlay_get(overlay_number)->pointer = value;
                    break;
                };
                case 0x03: {
                    // overlay enable status
                    overlay_get(overlay_number)->enabled = value != 0;
                    break;
                };
            }

            break;
        };

        case 0x80000400 ... 0x80000401: { // mouse port
            uint8_t setting = port & 0x000000FF;
            switch (setting) {
                case 0x00: {
                    // button states
                    mouse.clicked = value & 0b001;
                    mouse.released = value & 0b010;
                    mouse.held = value & 0b100;
                    break;
                };
                case 0x01: {
                    // position
                    mouse.x = value & 0x0000FFFF;
                    mouse.y = (value & 0xFFFF0000) >> 16;
                    break;
                };
            }

            break;
        };

        case 0x80000600 ... 0x80000680: { // audio port
            size_t id = port & 0xFF;
            uint8_t channel = (id & 0x30) >> 4;
            uint8_t reg = (id & 0x0F);
            switch (id) {
                case 0x80: {
                    // AUDBASE
                    snd.base = value;
                    break;
                }
            }
            switch (reg) {
                case 0x0: {
                    // AUDxSTART
                    snd.channel[channel].start = value;
                    break;
                }
                case 0x1: {
                    // AUDxEND
                    snd.channel[channel].end = value;
                    break;
                }
                case 0x2: {
                    // AUDxLOOPSTART
                    snd.channel[channel].loop_start = value;
                    break;
                }
                case 0x3: {
                    // AUDxLOOPEND
                    snd.channel[channel].loop_end = value;
                    break;
                }
                case 0x4: {
                    // AUDxRATE
                    snd.channel[channel].frequency = value;
                    break;
                }
                case 0x5: {
                    // AUDxCONTROL
                    snd.channel[channel].volume = value & 0x0000007F;
                    snd.channel[channel].loop = value & 0x00000080;
                    snd.channel[channel].enable = value & 0x00000100;
                    snd.channel[channel].bits16 = value & 0x00000200;
                    break;
                }
                case 0x6: {
                    // AUDxPAN
                    snd.channel[channel].right_volume = value & 0x000000FF;
                    snd.channel[channel].left_volume = (value & 0x0000FF00) >> 8;
                    break;
                }
            }
            break;
        }

        case 0x80001000 ... 0x80005003: { // disk controller port
            size_t id = port & 0xFF;
            uint8_t operation = (port & 0x0000F000) >> 8;
            switch (operation) {
                case 0x10: {
                    // no-op
                    break;
                };
                case 0x20: {
                    // set the buffer pointer
                    disk_controller.buffer_pointer = value;
                    break;
                };
                case 0x30: {
                    // read specified disk sector into memory
                    set_disk_sector(id, value);
                    read_disk_into_memory(id);
                    break;
                };
                case 0x40: {
                    // write specified disk sector from memory
                    set_disk_sector(id, value);
                    write_disk_from_memory(id);
                    break;
                };
                case 0x50: {
                    // remove specified disk
                    remove_disk(id);
                    break;
                };
            }

            break;
        };

        case 0x80010000: { // power control port
            if (value == 0) {
                bus_requests_exit = true;
            }
        };
    }

    return 0;
}

void drop_file(char *filename) {
    int last_id = 0;
    for (int i = 0; i < 4; i++) {
        if (disk_controller.disks[i].size != 0) {
            last_id++;
        }
    }
    new_disk(filename, last_id);
}
